### PR TITLE
Add RedirectLoginOptions to createAuthGuard

### PR DESCRIPTION
**Pull request by june07**
### Changes

Add authorizationParams option to authGuard to allow customization of auth behaviour based on routes. My particular use case required redirect_uri depending on which route was being guarded. Without this option there really wasn't a way to change the redirect_uri... or many of the other authorizationParams of the base auth0 client.

### Testing

- [ ] This change adds unit test coverage
- [ ] This change adds integration test coverage
- [x] This change has been tested on the latest version of the platform/language

### Checklist

- [x] I have read the [Auth0 general contribution guidelines](https://github.com/auth0/open-source-template/blob/master/GENERAL-CONTRIBUTING.md)
- [x] I have read the [Auth0 Code of Conduct](https://github.com/auth0/open-source-template/blob/master/CODE-OF-CONDUCT.md)
- [x] All code quality tools/guidelines have been run/followed

### CODE REVIEW

**Review by frederikprijck — CHANGES_REQUESTED** [review]
Thanks for the changes, i do think this is great, but needs a bit more consideration beyond your use case.

- the createGuardHandler needs to take RedirectLoginOptions, not just AuthorizationParams. As some peoole might want different options outside of authorizationParams.
- we need to ensure the authGuard keeps working as is with typescript, so we definetly should not add required arguments like being done. If we can, we should leave it optional, if not we should drop it and only support the options through the createGuardHandler.

We will also need tests for this new functionality.

Thanks for the contribution!

File: src/guard.ts
Comment (frederikprijck) [inline src/guard.ts:22]: Let's keep this as it was, and spread the options

```suggestion
    await client.loginWithRedirect({
      appState: { target: to.fullPath },
      ...redirectLoginOptions
    });
```

File: src/guard.ts
Comment (frederikprijck) [inline src/guard.ts:11]: We can drop this line.

File: src/guard.ts
Comment (frederikprijck) [inline src/guard.ts:45]: This breaks using the guard the way people should:

```
{
  path: '/profile',
  name: 'profile',
  component: Profile,
  beforeEnter: authGuard
}
```
We should ensure we do not break existing behavior.

**Review by frederikprijck — CHANGES_REQUESTED** [review]

File: src/guard.ts
Comment (june07) [reply src/guard.ts:43]: Not sure I follow on this one... the redirectLoginOptions is optional...

`redirectLoginOptions?:`

And the current api only shows authGuard having a "to" parameter. I'm assuming you're referring to when it's used in the above way and passed both "to" and "from" from Vue router? Still this api doesn't seem to do anything with the "from", and I'm not seeing how things break.

Are you saying that the function signature should be more like this:
authGuard(to: RouteLocation, from?: RouteLocation, redirectLoginOptions?: RedirectLoginOptions)

Sorry, but I'm unclear on this one...

Comment (frederikprijck) [reply src/guard.ts:42]: What I am saying is that our example we have (see playground/router.ts) no longer compiles:

image

It fails because Vue Router's Guard do not accept an optional RedirectLoginOptions. So we need to look into seeing if we can fix the types in such a way that Vue Router likes it, or remove the redirectLoginOptions alltogether

Comment (frederikprijck) [reply src/guard.ts:42]: > Are you saying that the function signature should be more like this:
> `authGuard(to: RouteLocation, from?: RouteLocation, redirectLoginOptions?: RedirectLoginOptions)`

No, that won't compile neither. 

I think we should have:

```ts
export async function authGuard(
  to: RouteLocation,
  from?: RouteLocation,
  next?: NavigationGuardNext,
  redirectLoginOptions?: RedirectLoginOptions
) {
  const auth0 = unref(auth0Client);

  return createGuardHandler(auth0, to, redirectLoginOptions);
}
```

Comment (frederikprijck) [reply src/guard.ts:42]: I know we do not use from and next, but a Guard in Vue is a function that takes 3 arguments, in a specific order: to, from and next. We can't just add `redirectLoginOptions` as a second argument.

Comment (frederikprijck) [reply src/guard.ts:42]: We will also need the exact same change to `createAuthGuard`.

Comment (june07) [reply src/guard.ts:42]: I got ya. I'll make the change in a bit and push an update. I didn't see the 3rd argument in the Vue docs per https://router.vuejs.org/guide/advanced/navigation-guards.html#optional-third-argument-next But I understand why this must be supported for backward compatibility.

Thank you for the clarification.

Comment (frederikprijck) [reply src/guard.ts:42]: No worries, thank you for the PR! 🚀

**Review by frederikprijck — CHANGES_REQUESTED** [review]
Thanks for all the changes. Is there any reason not to update createAuthGuard in the same way?

I believe we need to be consistent, and offer the same API for both.

**Comment by june07** [comment]
Reading back through your [commit here](https://github.com/auth0/auth0-vue/pull/80/commits/a2d16ffb130333b867939a3ec42a2b549b9fcbe3) and associated PR, it just seems that createAuthGuard by design shouldn't be used in the router components but rather closer to the app setup. Assuming that is the case, authGuard should be used if/when greater functionality provided by this PR is needed, and not createAuthGuard. To me it just makes things clearer as to when to use what in this API.

Specifically quoting directly from @maxweld in https://github.com/auth0/auth0-vue/issues/73...
> it requires the 'app' and the 'router' to be defined in the same component, or requires that the 'app' be passed to the router component

In my opinion, not changing createAuthGuard helps to maintain the clean separation you have previously made...

However, if I'm missing something, let me know and I'll certainly make the extra changes.

**Comment by frederikprijck** [comment]
Are you saying that anyone using `createAuthGuard` should not have the possibility to override the options that are being introduced in this PR, if so, why?

createAuthGuard is introduced to allow using a guard with multiple Vue apps, which shouldnt mean they can not override the RedirectLoginOptions.

**Review by frederikprijck — APPROVED** [review]
Thank you for all the work here 🚀 

I did a couple of patch commits, basically to remove some arguments, add missing tests and fix formatting.

**Comment by frederikprijck** [comment]
I do want to mention that we have an issue with our Browserstack environment with all our SDKs. We are actively working to get that resolved. Until then, we can't merge this PR but will work to get this merged and released as soon as possible!

**Comment by frederikprijck** [comment]
Hey @june07 ,

After further testing, this change does not work at all. Did you verify this in your own project or for your own use-case?

Vue-Router does not like the fact that we added the `next` argument, but didn't use it. Therefore we will need to provide this functionality in a different way:

- Keep the next argument, and replace returning a value from the Guard with calling next with the same value.
- Remove the next argument again, but only support RedirectLoginOptions through `createAuthGuard`

The first one will be a breaking change, which I would prefer not to do when we can work around it differently.

I added a commit that should allow achieving what you want, but a bit differently without causing a breaking change for existing users.

**Comment by june07** [comment]
> After further testing, this change does not work at all. Did you verify this in your own project or for your own use-case?
Yes, BEFORE the requested changes (next argument) it works beautifully, yes. I am still using the npm package from my original PR with success in my own project.

Question: shouldn't the test [here](https://github.com/auth0/auth0-vue/blob/e719d629c27fb0208513655f6d6cca1be4a01373/playground/router.ts#L28) be executing authGuard?

As I am using authGuard in another project...



`beforeEnter: (to) => authGuard(to, { screen_hint: 'signup', redirect_uri: window.location.origin + '/?returnTo=synchook' })`

However, the test above is simply passing authGuard...

**Comment by frederikprijck** [comment]
Yes, that uses the route guard and resulted in an error in the console when you try and go to that route.

```
beforeEnter: (to) => authGuard(to, { screen_hint: 'signup', redirect_uri: window.location.origin + '/?returnTo=synchook' })
```

Not sure I understand this, shouldnt this also include `from` and `next`? As that does not reflect what was added in this PR.

```
beforeEnter: (to, from, next) => authGuard(to, from, next, { screen_hint: 'signup', redirect_uri: window.location.origin + '/?returnTo=synchook' })
```

With this PR (prior to my changes), that is what you wanted to be doing, but yes that won't catch the error I mean. 

The error can be seen here: https://github.com/vuejs/router/blob/c1f7cd769e7194a14092c47b2638482744d7d021/packages/router/src/navigationGuards.ts#L192-L195, and comes down to the fact that you can not have `next` without calling it.


The gist here is that before this PR, people could do
```
beforeEnter: authGuard
```
And they need to be able to keep doing this, but they can not prior to my last commit. And neither can they with the AuthGuard you are using in your project.

**Comment by frederikprijck** [comment]
Pushed a commit to support the following syntax:

```ts
createAuthGuard();
createAuthGuard(vueApp);
createAuthGuard({ app: vueApp });
createAuthGuard({ app: vueApp, redirectLoginOptions: { authorizationParams: {} } });
createAuthGuard({ redirectLoginOptions: { authorizationParams: {} } });
```

This way we can avoid breaking existing code, while also avoiding to have to set `app` to undefined when all you want is to set the redirectLoginOptions.

**Review by ewanharris — APPROVED** [review]

**Pull request merged**
